android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudApiTypes.java

```diff
@@ -0,0 +1,195 @@
+package com.tavultesoft.kmea.data;
+
+import android.app.DownloadManager;
+
+import androidx.annotation.NonNull;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+public class CloudApiTypes {
+  protected static class CloudApiReturns {
+
+    public final ApiTarget target;
+    public final JSONArray jsonArray;
+    public final JSONObject jsonObject;
+
+    public CloudApiReturns(ApiTarget target, JSONArray jsonArray) {
+      this.target = target;
+      this.jsonArray = jsonArray;
+      this.jsonObject = null;
+    }
+
+    public CloudApiReturns(ApiTarget target, JSONObject jsonObject) {
+      this.target = target;
+      this.jsonArray = null;
+      this.jsonObject = jsonObject;
+    }
+
+  }
+
+  protected enum ApiTarget {
+    Keyboards,
+    LexicalModels
+  }
+
+  protected enum JSONType {
+    Array,
+    Object
+  }
+
+  protected static class CloudApiParam {
+    public final ApiTarget target;
+    public final String url;
+    public final JSONType type;
+
+    CloudApiParam(ApiTarget target, String url, JSONType type) {
+      this.target = target;
+      this.url = url;
+      this.type = type;
+    }
+  }
+
+  public static class SingleCloudDownload
```

Comment:
I'm also having trouble trying to make sense of this class and how it fits into the download abstraction that seems to be going on.  Maybe some additional documentation on the general design and how each new class accomplishes that might help?

Reply:
For catalog update we need 2 physical downloads (SingleDownload), which has to be applied to the dataset in one final processing step.
You register all downloads belonging to one update process and the last one finishing starts the processing step.

Reply:
It will be easier to checkout the branch and see the CloudDonwloadMgr to understand how it works.
The most other changes are Refactoring the old Clourepository to be able to reuse the implementation and have both DonwnloadTask and DownloadManager